File: frontend/types/skill-gap.ts
/**
 * Mirrors backend/app/schemas/skill_gap.py exactly -- field-for-field,
 * same nullability. readiness_percentage/gap are plain Pydantic ints (not
 * Decimal), so -- unlike types/assessment.ts -- these come back as real
 * JSON numbers, not strings; still never recomputed client-side.
 */

import type { Difficulty } from "@/types/assessment";

export type Importance = "CORE" | "IMPORTANT" | "OPTIONAL";
export type RelationshipType = "PREREQUISITE" | "RELATED" | "NEXT_STEP" | "COMPLEMENTARY";
export type GapStatus = "MATCHED" | "NEEDS_IMPROVEMENT" | "MISSING";
export type VerificationStatus = "VERIFIED" | "UNVERIFIED";
export type Priority = "HIGH" | "MEDIUM" | "LOW";
export type AnalysisMode = "JOB_ROLE" | "PERSONAL";

/** Mirrors `JobRoleResponse`. Only ever populated from active roles. */
export interface JobRole {
  id: string;
  name: string;
  description: string | null;
  category: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

/** Mirrors `TargetJobRoleResponse`. */
export interface TargetJobRole {
  id: string;
  job_role: JobRole;
  created_at: string;
  updated_at: string;
}

/** Mirrors `SkillGapItem`. current_level is the student's DECLARED level
 * (or null if the skill isn't in their active list) -- verification_status
 * is a separate signal for whether that declared level is
 * assessment-verified. Never derive one from the other client-side. */
export interface SkillGapItem {
  skill_id: string;
  skill_name: string;
  current_level: Difficulty | null;
  required_level: Difficulty;
  gap: number;
  status: GapStatus;
  verification_status: VerificationStatus;
  importance: Importance;
  priority: Priority;
  assessment_available: boolean;
  assessment_id: string | null;
}

/** Mirrors `SkillGapSummary`. */
export interface SkillGapSummary {
  matched: number;
  needs_improvement: number;
  missing: number;
  unverified: number;
}

/** Mirrors `Recommendation`. `reason` is server-authored from a small set
 * of fixed templates -- never LLM text, never rewritten client-side. */
export interface Recommendation {
  skill_id: string;
  skill_name: string;
  reason: string;
  current_level: Difficulty | null;
  target_level: Difficulty | null;
  gap: number | null;
  priority: Priority;
  relationship_type: RelationshipType | null;
  is_missing: boolean;
  is_verified: boolean;
  assessment_available: boolean;
  assessment_id: string | null;
}

/** Mirrors `SkillGapJobRoleResponse` -- GET /skill-gap when a target role
 * is set, or GET /skill-gap/job-role/{id} for any active role. */
export interface SkillGapJobRoleAnalysis {
  mode: "JOB_ROLE";
  job_role: JobRole;
  readiness_percentage: number;
  summary: SkillGapSummary;
  skills: SkillGapItem[];
  recommendations: Recommendation[];
}

/** Mirrors `PersonalSkillCounts`. */
export interface PersonalSkillCounts {
  total_active_skills: number;
  verified_skills: number;
  unverified_skills: number;
  beginner_skills: number;
  intermediate_skills: number;
  advanced_skills: number;
  expert_skills: number;
}

/** Mirrors `ProgressableSkill`. */
export interface ProgressableSkill {
  skill_id: string;
  skill_name: string;
  current_level: Difficulty;
  next_level: Difficulty;
  assessment_available: boolean;
  assessment_id: string | null;
}

/** Mirrors `PrerequisiteGap`. */
export interface PrerequisiteGap {
  skill_id: string;
  skill_name: string;
  required_for_skill_id: string;
  required_for_skill_name: string;
}

/** Mirrors `SkillGapPersonalResponse` -- GET /skill-gap when no target
 * role is set. */
export interface SkillGapPersonalAnalysis {
  mode: "PERSONAL";
  counts: PersonalSkillCounts;
  progressable_skills: ProgressableSkill[];
  recommendations: Recommendation[];
  prerequisite_gaps: PrerequisiteGap[];
}

/** GET /skill-gap's response is one of these two shapes -- `mode`
 * discriminates which. Never render one without checking `mode` first. */
export type SkillGapAnalysis = SkillGapJobRoleAnalysis | SkillGapPersonalAnalysis;

File: backend/app/schemas/skill_gap.py
"""Pydantic schemas for Skill Gap Analysis, job roles, and the student's
target role.

Mirrors database/migrations/016_skill_gap.sql exactly -- no invented
columns, no invented enum values. Reuses the EXISTING four-value
proficiency scale from student_skills.proficiency_level /
assessments.difficulty (Beginner/Intermediate/Advanced/Expert) -- this
module defines no second, competing proficiency enum.

No LLM, no free-text AI-generated fields anywhere in this module. Every
"reason" string is built from a small set of deterministic templates in
app.services.skill_gap_service, driven entirely by job_role_skills /
skill_relationships / student_skills / assessments data.
"""

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.schemas.assessment import Difficulty

# ============================================================
# Enums
# ============================================================


class Importance(str, Enum):
    CORE = "CORE"
    IMPORTANT = "IMPORTANT"
    OPTIONAL = "OPTIONAL"


class RelationshipType(str, Enum):
    PREREQUISITE = "PREREQUISITE"
    RELATED = "RELATED"
    NEXT_STEP = "NEXT_STEP"
    COMPLEMENTARY = "COMPLEMENTARY"


class GapStatus(str, Enum):
    MATCHED = "MATCHED"
    NEEDS_IMPROVEMENT = "NEEDS_IMPROVEMENT"
    MISSING = "MISSING"


class VerificationStatus(str, Enum):
    VERIFIED = "VERIFIED"
    UNVERIFIED = "UNVERIFIED"


class Priority(str, Enum):
    HIGH = "HIGH"
    MEDIUM = "MEDIUM"
    LOW = "LOW"


class AnalysisMode(str, Enum):
    JOB_ROLE = "JOB_ROLE"
    PERSONAL = "PERSONAL"


# ============================================================
# job_roles / job_role_skills
# ============================================================


class JobRoleResponse(BaseModel):
    """Mirrors `job_roles`. Only ever populated from rows RLS already
    filtered to is_active = true."""

    id: UUID
    name: str
    description: str | None
    category: str | None
    is_active: bool
    created_at: datetime
    updated_at: datetime


class JobRoleListResponse(BaseModel):
    job_roles: list[JobRoleResponse]


class JobRoleSkillRequirement(BaseModel):
    """Mirrors `job_role_skills`, with the skill's name/category embedded
    for display -- never just a bare skill_id."""

    skill_id: UUID
    skill_name: str
    category_name: str | None
    required_level: Difficulty
    importance: Importance


class JobRoleDetailResponse(BaseModel):
    role: JobRoleResponse
    requirements: list[JobRoleSkillRequirement]


# ============================================================
# student_target_job_role
# ============================================================


class TargetJobRoleResponse(BaseModel):
    """Mirrors `student_target_job_role`, joined with the role itself.
    student_id is deliberately NOT included -- this is always the
    caller's own row, resolved server-side; there is nothing to gain by
    echoing it back."""

    id: UUID
    job_role: JobRoleResponse
    created_at: datetime
    updated_at: datetime


class SetTargetJobRoleRequest(BaseModel):
    """The only field a client may ever supply -- student_id is derived
    from the authenticated session, never accepted here."""

    model_config = ConfigDict(extra="forbid")

    job_role_id: UUID


# ============================================================
# Skill gap analysis
# ============================================================


class SkillGapItem(BaseModel):
    """One required skill's gap status against the target role.

    current_level is the student's DECLARED student_skills.proficiency_level
    (or None if the skill isn't in their active list at all) --
    verification_status is a SEPARATE signal for whether that declared
    level has actually been assessment-verified. These are never mixed
    into one value (see the module/service docstrings for why).
    """

    skill_id: UUID
    skill_name: str
    current_level: Difficulty | None
    required_level: Difficulty
    gap: int
    status: GapStatus
    verification_status: VerificationStatus
    importance: Importance
    priority: Priority
    assessment_available: bool
    assessment_id: UUID | None


class SkillGapSummary(BaseModel):
    matched: int
    needs_improvement: int
    missing: int
    unverified: int


class Recommendation(BaseModel):
    """One deterministic, data-driven recommendation -- never LLM-authored
    free text. `reason` is built from a small fixed set of templates
    (see skill_gap_service.py) parameterized only by real data (skill
    names, levels, relationship type)."""

    skill_id: UUID
    skill_name: str
    reason: str
    current_level: Difficulty | None
    target_level: Difficulty | None
    gap: int | None
    priority: Priority
    relationship_type: RelationshipType | None
    is_missing: bool
    is_verified: bool
    assessment_available: bool
    assessment_id: UUID | None


class SkillGapJobRoleResponse(BaseModel):
    mode: AnalysisMode
    job_role: JobRoleResponse
    readiness_percentage: int
    summary: SkillGapSummary
    skills: list[SkillGapItem]
    recommendations: list[Recommendation]


# ---- Personal (no target role) mode ----


class PersonalSkillCounts(BaseModel):
    total_active_skills: int
    verified_skills: int
    unverified_skills: int
    beginner_skills: int
    intermediate_skills: int
    advanced_skills: int
    expert_skills: int


class ProgressableSkill(BaseModel):
    """One of the student's current skills that has a real assessment
    available at the next level up."""

    skill_id: UUID
    skill_name: str
    current_level: Difficulty
    next_level: Difficulty
    assessment_available: bool
    assessment_id: UUID | None


class PrerequisiteGap(BaseModel):
    """A skill the student is missing that is a documented PREREQUISITE
    (skill_relationships) of one of their recommended next skills."""

    skill_id: UUID
    skill_name: str
    required_for_skill_id: UUID
    required_for_skill_name: str


class SkillGapPersonalResponse(BaseModel):
    mode: AnalysisMode
    counts: PersonalSkillCounts
    progressable_skills: list[ProgressableSkill]
    recommendations: list[Recommendation]
    prerequisite_gaps: list[PrerequisiteGap]
